File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.cpp
/**
 * 1. length-n sorted array (asending order)
 * 2. rotated 1~n times: n times -> unchanged
 * 3. all elements are unique
 * 4. O(log n) solution
 * Given: rotated sorted array ``nums`` and an ``target``.
 * Find: index of ``target`` within ``nums``.
 */

 /**
  * O(log n) Solution: bineary search vriation
  * `[3, 4, 5, 1, 2]` & `4` -> first half
  * `[3, 4, 5, 1, 2]` $ `1` -> second half
  * `[4, 5, 1, 2, 3]` $
  */

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                if (nums[left] == target) {
                    return left;
                } else if (nums[left] < target) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[right] == target) {
                    return right;
                } else if (nums[right] > target) {
                    left = mid + 1;
                } else {
                    right = mid -1;
                }
            }

        }
        return -1;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        """
        - Relation: sorted
        - Complexity: O(log n) time
        Binary Search!
        """
        l, r = 0, len(nums) - 1

        while l <= r: # The final position is possible candidate
            m = l + (r - l) // 2 # Prevent overflow
            # nums[m] is checked, no check again
            if nums[m] == target:
                return m
            
            # 1. Left half is sorted
            # [Q] `<` or `<=`? [A] no double check, so `<`
            if nums[l] < nums[m]:
                # target in the sorted part
                if target < nums[m] and nums[l] <= target:
                    r = m - 1
                else:
                    l = m + 1

            # 2. Right half is sorted
            else:
                # target in the sorted part
                if target > nums[m] and nums[r] >= target:
                    l = m + 1
                else:
                    r = m - 1

        return -1
            